Support Mach-O backtraces without dsymutil

philipc opened the pull request:
This is currently all horrible code, and it doesn't cache anything, but it passes the tests. It depends on some unmerged changes in `object`, so I'll get them cleaned up and merged first.

alexcrichton commented on src/symbolize/gimli/macho.rs:
```diff
@@ -201,4 +206,51 @@ impl<'a> Object<'a> {
+
+fn split(path: &str) -> Option<(&str, &str)> {
+    let index = path.find('(')?;
+    let (archive, rest) = path.split_at(index);
```
oh wow it's literally `libfoo.a(foo.o)`!

alexcrichton reviewed (commented):
Nice!

I think that this looks great and would be happy to land it feature gated. For caching, I wonder if we could perhaps just have an infinitely-sized "cache" inside of each object? That way the LRU can take care of top-level things, but each top-level thing is allowed to be arbitrarily big (which sorta makes sense to me). Basically each `Object` in `backtrace` would have a `Vec` of loaded member objects, populated lazily of course.

philipc commented:
> I think that this looks great and would be happy to land it feature gated.

What do you think it would take to not need to be gated?

> For caching, I wonder if we could perhaps just have an infinitely-sized "cache" inside of each object? That way the LRU can take care of top-level things, but each top-level thing is allowed to be arbitrarily big (which sorta makes sense to me). Basically each Object in backtrace would have a Vec of loaded member objects, populated lazily of course.

Yeah, that's the kind of thing I was expecting. By "infinitely-sized", were you expecting that we'd do a linear scan for the object file name? Instead, since we can already identify each object file by an index, I've used that to index into a presized Vec. It doesn't try to share mmaps for archives though. I was actually wondering if the OS would be smart enough to make that cheap.

alexcrichton commented:
> What do you think it would take to not need to be gated?

Oh not much probably! If you'd prefer to land it ungated I don't mind reviewing accordingly. I think though it's nothing really out of the ordinary, just a thorough review, tests, CI, etc.

> Yeah, that's the kind of thing I was expecting. By "infinitely-sized", were you expecting that we'd do a linear scan for the object file name?

Would it be possible to have some sort of indexed-by-address map of some sort? I'm not really sure how this debuginfo stuff works, but can we always say that everything within a range of addresses will lie in the same object file? Ideally we could have a `Vec` of objects which is sorted by the addresses they show up at, and then we can just lookup an address in this list first to find the dwarf info for a symbol. Failing that we could try the main file or loading the object lazily.

philipc commented:
> Oh not much probably! If you'd prefer to land it ungated I don't mind reviewing accordingly. I think though it's nothing really out of the ordinary, just a thorough review, tests, CI, etc.

I think that should be doable, but hold off until I get CI passing (this needs a least new `object` release). I assume the existing tests with `RUSTFLAGS="-Z run-dsymutil=no"` will be enough?

> Would it be possible to have some sort of indexed-by-address map of some sort? I'm not really sure how this debuginfo stuff works, but can we always say that everything within a range of addresses will lie in the same object file? Ideally we could have a Vec of objects which is sorted by the addresses they show up at, and then we can just lookup an address in this list first to find the dwarf info for a symbol. Failing that we could try the main file or loading the object lazily.

I don't think there is any guarantee that each object will be a contiguous address range, so we can't exactly do that. The next best thing is a Vec of symbols sorted by addresses, and we already have that in `object::ObjectMap`. So we do a binary search for the address which gives us an object index, then use the object index to directly get the dwarf info for that object, then look up the address in the dwarf info.

Another reason we need it to be a Vec of symbols is that the addresses in the dwarf haven't been relocated, so `object::ObjectMap` also contains the symbol names, which lets us look up the original symbol in the object file to "unrelocate" the address  (currently this is a linear scan, but could be changed to a binary search or a hash map).

alexcrichton commented:
For testing I think yeah `-Zrun-dsymutil=no` should probably be good enough. I at least don't know how to test much more myself!

Given what you're saying, could perhaps each object file internally have a list of dwarf debug sections, and we have a mapping from the object index to the position in the vector? That way if it's in the original file itself it'll be at index 0, and otherwise later indexes are populated by lazily loading the object indices.

I'm not sure if that makes sense, but I can try pushing up a commit with what I'm thinking if it doesn't.

alexcrichton commented:
Er carry on, what this PR does now is exactly what I was thinking 👍

alexcrichton commented on src/symbolize/gimli.rs:
```diff
@@ -613,7 +613,18 @@ pub unsafe fn resolve(what: ResolveWhat<'_>, cb: &mut dyn FnMut(&super::Symbol))
             }
         }
-
+        if let Some((object_cx, object_addr)) = cx.object.search_object_map(addr as u64) {
```
I played around with this a bit locally, ideally we'd deduplciate this and the above `find_frames` loop since I think they're basically identical. Unfortunately I kept running into lifetime errors. It may mean that the lazily-populated map for objects should perhaps use interior mutability, or ideally something like `LazyCell`, although that's not stable yet.

philipc replied:
I wasn't able to come up with anything for this that felt like a net decrease in complexity.

alexcrichton reviewed (commented):
I left a few comments here and there. I think https://github.com/rust-lang/backtrace-rs/pull/379 has some cleanup which may make some pieces nicer here eventually, but not as much as I thought so I'll wait on this merging before that.

philipc commented on src/symbolize/gimli/macho.rs:
```diff
@@ -201,4 +213,72 @@ impl<'a> Object<'a> {
+                let object = Object::parse(macho, endian, data)?;
+                let stash = Stash::new();
+                let inner = super::cx(&stash, object)?;
+                return Some(mk!(Mapping { map, inner, stash }));
```
This is already deduplicated in your other comment, right?

philipc commented:
I've rebased and force pushed. I think I addressed most of the comments, and I also changed the object symbol lookup to a binary search.

alexcrichton reviewed (commented):
Ok sounds good to me! A last thing or two but otherwise looks good-to-go to me!

bjorn3 commented on src/symbolize/gimli/macho.rs:
```diff
@@ -137,21 +137,29 @@ fn find_header(mut data: Bytes<'_>) -> Option<(&'_ Mach, Bytes<'_>)> {
         while let Ok(Some(command)) = commands.next() {
             if let Some((segment, section_data)) = MachSegment::from_command(command).ok()? {
-                if segment.name() == b"__DWARF" {
+                if segment.name() == b"__DWARF" || (is_object && segment.name() == b"") {
```
This is the segment name, not section name. The section name is always empty in object files. Only linked dylibs and executables use segments.

alexcrichton replied:
Ah right makes sense!

philipc replied:
Or to be pedantic, the individual sections do still specify the correct segment name in the section struct, but all of the sections in an object file are placed in a single unnamed segment load command.

philipc commented:
I added those comments and the debug asserts for the symbol sorting. Two more things
- the macos-nightly no-dsymutil test wasn't testing anything because it was still finding the dsym it had already built. So I added a `cargo clean`, but maybe we should be checking timestamps or something on these?
- I don't know why the windows-gnu32 build is failing now, but I checked that the rest of the tests still pass with it deleted.

alexcrichton commented:
Ah yeah I think that's anomalous, so I'm going to go ahead and merge, thanks again @philipc!
